tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/JavaScope.java

```diff
@@ -0,0 +1,153 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow.op;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.tensorflow.DeviceSpec;
+import org.tensorflow.ExecutionEnvironment;
+import org.tensorflow.Operation;
+import org.tensorflow.OperationBuilder;
+
+/**
+ * A Java implementation of {@link Scope}. This is used in all cases except custom gradient
+ * definitions.
+ */
+public final class JavaScope implements Scope {
```

Comment:
What about renaming this class to `OpScope` and `NativeScope` to `GradientScope`?